// @centrifuge/sdk/dist/entities/IndexerQueries/assetTransactions.js
import { Balance, Price } from '../../utils/BigInt.js';
export var AssetType;
(function (AssetType) {
    AssetType["OnchainCash"] = "OnchainCash";
    AssetType["OffchainCash"] = "OffchainCash";
    AssetType["Other"] = "Other";
})(AssetType || (AssetType = {}));
export const assetTransactionsPostProcess = (data) => {
    return (data.assetTransactions.nodes.map((tx) => {
        const decimals = tx.pool.currency.decimals;
        return {
            ...tx,
            name: tx.asset.name,
            settlementPrice: tx.settlementPrice ? new Price(tx.settlementPrice) : null,
            amount: new Balance(tx?.amount ?? 0n, decimals),
            principalAmount: tx.principalAmount ? new Balance(tx.principalAmount, decimals) : null,
            interestAmount: tx.interestAmount ? new Balance(tx.interestAmount, decimals) : null,
            realizedProfitFifo: tx.realizedProfitFifo ? new Balance(tx.realizedProfitFifo, decimals) : undefined,
            sumRealizedProfitFifo: tx.asset.sumRealizedProfitFifo
                ? new Balance(tx.asset.sumRealizedProfitFifo, decimals)
                : undefined,
            unrealizedProfitAtMarketPrice: tx.asset.unrealizedProfitAtMarketPrice
                ? new Balance(tx.asset.unrealizedProfitAtMarketPrice, decimals)
                : undefined,
            timestamp: new Date(`${tx.timestamp}+00:00`),
        };
    }) || []);
};
export const assetTransactionsQuery = `
query($filter: AssetTransactionFilter) {
  assetTransactions(
    orderBy: TIMESTAMP_ASC,
    filter: $filter
  ) {
    nodes {
      principalAmount
      interestAmount
      epochId
      type
      timestamp
      amount
      settlementPrice
      quantity
      hash
      realizedProfitFifo
      pool {
        currency {
          decimals
        }
      }
      asset {
        id
        metadata
        name
        type
        sumRealizedProfitFifo
        unrealizedProfitAtMarketPrice
      }
      fromAsset {
        id
        metadata
        name
        type
      }
      toAsset {
        id
        metadata
        name
        type
      }
    }
  }
}
`;
