// @centrifuge/sdk/dist/entities/IndexerQueries/poolFeeSnapshots.js
import { Balance } from '../../utils/BigInt.js';
export function poolFeeSnapshotsPostProcess(data) {
    const poolFeeSnapshots = data.poolFeeSnapshots;
    const poolFeesGroupedByDate = poolFeeSnapshots?.nodes.reduce((acc, snapshot) => {
        const date = snapshot.timestamp.slice(0, 10);
        if (!acc[date]) {
            acc[date] = [];
        }
        const poolCurrencyDecimals = snapshot.poolFee.pool.currency.decimals;
        const poolFeeSnapshot = {
            timestamp: snapshot.timestamp,
            pendingAmount: new Balance(snapshot.pendingAmount, poolCurrencyDecimals),
            poolFee: {
                name: snapshot.poolFee.name,
            },
            poolFeeId: snapshot.poolFeeId,
            poolCurrency: {
                decimals: poolCurrencyDecimals,
            },
            sumAccruedAmount: new Balance(snapshot.sumAccruedAmount, poolCurrencyDecimals),
            sumChargedAmount: new Balance(snapshot.sumChargedAmount, poolCurrencyDecimals),
            sumPaidAmount: new Balance(snapshot.sumPaidAmount, poolCurrencyDecimals),
            sumAccruedAmountByPeriod: new Balance(snapshot.sumAccruedAmountByPeriod, poolCurrencyDecimals),
            sumChargedAmountByPeriod: new Balance(snapshot.sumChargedAmountByPeriod, poolCurrencyDecimals),
            sumPaidAmountByPeriod: new Balance(snapshot.sumPaidAmountByPeriod, poolCurrencyDecimals),
        };
        acc[date].push(poolFeeSnapshot);
        return acc;
    }, {});
    return poolFeesGroupedByDate;
}
export const poolFeeSnapshotQuery = `
query($filter: PoolFeeSnapshotFilter) {
  poolFeeSnapshots(
    orderBy: BLOCK_NUMBER_ASC, 
    filter: $filter
  ) {
    nodes {
        id
        poolFeeId
        timestamp
        sumPaidAmount
        sumChargedAmount
        sumAccruedAmount
        sumPaidAmount
        pendingAmount
        sumAccruedAmountByPeriod
        sumPaidAmountByPeriod
        sumChargedAmountByPeriod
        poolFee {
          name
          pool {
            currency {
              decimals
            }
          }
        }
    }
  }
}
`;
