// @centrifuge/sdk/dist/tests/mocks/mockInvestorTransactions.js
import { Price } from '../../utils/BigInt.js';
import { Balance } from '../../utils/BigInt.js';
export const mockInvestorTransactions = [
    {
        id: 'tx-1',
        poolId: 'pool-1',
        timestamp: new Date('2024-01-01T12:00:00Z'),
        accountId: 'account-1',
        chainId: 1,
        evmAddress: '0x123a',
        trancheId: 'senior',
        epochNumber: 1,
        type: 'INVEST_ORDER_UPDATE',
        currencyAmount: new Balance(1000000n, 6), // 1.0
        tokenAmount: new Balance(900000n, 6), // 0.9
        tokenPrice: new Price(1100000000000000000n), // 1.1
        hash: '0xabc',
    },
    {
        id: 'tx-2',
        poolId: 'pool-1',
        timestamp: new Date('2024-01-01T18:00:00Z'),
        accountId: 'account-1',
        chainId: 1,
        evmAddress: '0x123b',
        trancheId: 'senior',
        epochNumber: 1,
        type: 'INVEST_EXECUTION',
        currencyAmount: new Balance(2000000n, 6), // 2.0
        tokenAmount: new Balance(1800000n, 6), // 1.8
        tokenPrice: new Price(1100000000000000000n), // 1.1
        hash: '0xdef',
    },
];
